### DISCUSSION TITLE
Tore calibration and error margin depending on calibration value

### DISCUSSION THREAD

**Discussion by mathieucarbou** [OpenEVSE/open_evse#23] [Q&A] [answered]
Hello,

Default calibration for ampmeter coming with OpenEVSE box is `$SA 220 0`. Which is perfect for normal / high charge current.

But when using solat diver, lower charge current are completely of by 0.25A and I need to set `$SA 190 0` to calibrate.

Here is a comparison with a Shelly EM which was confirmed being accurate:

![image](https://user-images.githubusercontent.com/61346/236951149-f269845f-d030-47c1-8a5f-037237a407da.png)

is there a way to set a sort of factor so that 190 is used for lower current and then 220 ?

Right now, in Home Assistant I am using an automation that is setting the right calibration value depending if I am using AUTO mode with solar divert (which means low charge currents) or Manual mode (which mean high charge current)

```yaml
  - alias: "OpenEVSE: Bugfix Calibration"
    trigger:
      - platform: state
        entity_id:
          - select.openevse_charge_control
          - switch.openevse_solar_divert
    condition: []
    action:
      - if:
          - condition: state
            entity_id: switch.openevse_solar_divert
            state: "on"
          - condition: state
            entity_id: select.openevse_charge_control
            state: Auto
        then:
          - service: mqtt.publish
            data:
              topic: "openevse/rapi/in/$SA"
              payload: "190 0"
        else:
          - service: mqtt.publish
            data:
              topic: "openevse/rapi/in/$SA"
              payload: "220 0"
    mode: single
```

**Comment by mathieucarbou** [comment]
@KipK : according to your comment here (https://github.com/OpenEVSE/ESP32_WiFi_V4.x/pull/634#issuecomment-1544457609) RAPi support will be removed from MQTT.

So how to fix this issue ?

The tore, even if calibrated, leads to wrong results because its calibration seems to be based on a certain power range. It is not accurate both for low current (like 6-7A in divert eco mode) and 20A (in normal mode)

**Comment by KipK** [comment, marked as answer]
Are you sure the difference you get is not all about power losses?
Those amp sensor clamps has less than 3% bias and are quiet liners so I doubt you have to use different calibration depending of current. 
However if your Shelly is at the beginning of the cables, you miss the power loss crossing those cables to the evse. 

I see more efficiency by charging high current than low ones.

**Reply by mathieucarbou** [reply]
> Are you sure the difference you get is not all about power losses? Those amp sensor clamps has less than 3% bias and are quiet liners so I doubt you have to use different calibration depending of current. However if your Shelly is at the beginning of the cables, you miss the power loss crossing those cables to the evse.
> 
> I see more efficiency by charging high current than low ones.

Oh yes maybe! Let's see... The shelly is indeed at the beginning of the cables. I am charing at 18A max (real current) and I have I have:

- 10m of 3G6 to the box, so -1.38V (0.6%)
- 10m 4G2.5 between the box and the T2 socket, so -3.31V (1.44%)

This is a global loss of 4.69V (2%).

If we look at the graph, the correct calibration for me would be `$SA 190 0`. I think

At 4000W seen by the shelly, 3200W (80%) should go through the box. It seems to only see less, like only the loss by the 10m 3G2.5 cable though.

Thanks for the help!

**Comment by mathieucarbou** [comment]
@KipK : is there a possibility to have a voltmeter in the box to measure te voltage drop ? Right now, I am feeding the voltage through mqtt, but I am feeding the voltage seen by the shelly. Ideally, the box would need to read the voltage right ?

**Reply by KipK** [reply]
Yes look at  openevse git here , there's a PR for those cheap ZMPT boards :

https://github.com/OpenEVSE/open_evse/pull/13

**Comment by mathieucarbou** [comment]
Found an alternative:

I am computing my cable loss constant:

https://schema-electrique.net/calcul-chute-de-tension-electrique-formule-calcul-section-cable-triphase-monophase.html

`loss = 2×0.023×10/2.5 + 2×0.023×10/6`

Then, i am sending to the mqtt the grid voltage reduced :

` sent_voltage == grid_voltage - (loss * grid_power / grid_voltage)`

grid_power and grid_voltage are coming from a shelly EM.

Doing that, I can keep the default calibration to `$SA 220 0` and the current value computed by the evse module is exactly the same as `grid_power / grid_voltage`. So only the voltage decreases (consequently the evse session power).

**Comment by mathieucarbou** [comment]
@KipK : I am wondering if the "power" reported by OpenEVSE is in reality not real power (W) but instead apparent power (VA).

I am trying to fine-tune the calibration with the help of an OBD device, which reports the power, volts, current as seen by the car.

My Shelly EM is reporting power, power factor and voltage. I can derive the amperage by doing A = W / PF / V, which gives accurately the same amperage as seen by the car (but the car of course sees a lower voltage due to the cable loss).

So now, I am injecting to mqtt a voltage that is taking in consideration the cable loss based on the current from the shelly to the evse. The value is also accurate with the reading in the car (reduced by the loss from the box to the car).

**But the ammeter is measuring the current, so can he get the real difference between active and reactive power or not ?**

Currently, the power reported, is higher than the reported power of the shelly. Could it be because the power reported is actually VA and not Watts: if I consider the power factor, then the value matches.

**Reply by KipK** [reply]
For AC EV Charger, 99% of the current consumed is that consumed directly by the vehicle.
The power factor on evse is almost 1, but not the case of the EV inverter.

**Reply by mathieucarbou** [reply]
> For AC EV Charger, 99% of the current consumed is that consumed directly by the vehicle.
> The power factor on evse is almost 1, but not the case of the EV inverter.

I am measuring it with a Shelly EM and for low current, it is around 0.8, and for higher current when the PFC triggers it goes up to 0.98 (for my Zoe)

Shelly EM being on the phase going to the evse box.

That is why I have differences at low current between the car report (OBD), the ammeter from the evse and the shelly, even when accounting for the voltage loss. If the ammeter is not capable of seeing this power factor, it reports a higher power at low current like 6A than the Shelly EM. At high currents, they report the same value (after accounting for a little voltage loss)

**Reply by KipK** [reply]
How can the Shelly knows the power factor without knowing the hardware in the loop ?

**Reply by mathieucarbou** [reply]
> How can the Shelly knows the power factor without knowing the hardware in the loop ?

It is my understanding that the Shelly EM is capable of measuring the current shifts (how - I don't specifically know, but I know it has to be powered with the same phase). A shelly EM is capable of measuring the active power and reactive power, which means capable of detecting the shifts between current and voltage waves. The fact that is can take measurements at both points (its own powering current and the CT clamp might help).

![image](https://github.com/OpenEVSE/open_evse/assets/61346/b6f886d1-a559-4259-99ec-1be53ddc6874)

**Comment by chris1howell** [comment]
Calibration can be done with both Scale and Offset values.

Pick a current in the middle of your desired range and calibrate Scale for that value. Then modify the offset value so that both the Low and High end are close.

**Reply by mathieucarbou** [reply]
thanks! I was able to achieve a better calibration with this hint.
